pkg/utils/nodeclaim/nodeclaim.go

```diff
@@ -433,3 +434,113 @@ func CheckNodeClass(ctx context.Context, kClient client.Client) error {
 	}
 	return nil
 }
+
+// GetBringYourOwnNodes finds all BYO nodes that match the workspace's label selector
+func GetBringYourOwnNodes(ctx context.Context, c client.Client, wObj *kaitov1beta1.Workspace) ([]*v1.Node, error) {
+	// List all nodes in the cluster
+	nodeList := &v1.NodeList{}
+	listOpts := []client.ListOption{}
+
+	// If there's a label selector, add it to the list options
+	if wObj.Resource.LabelSelector != nil {
+		selector, err := metav1.LabelSelectorAsSelector(wObj.Resource.LabelSelector)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert label selector: %w", err)
+		}
+		listOpts = append(listOpts, client.MatchingLabelsSelector{Selector: selector})
+	}
+
+	if err := c.List(ctx, nodeList, listOpts...); err != nil {
+		return nil, fmt.Errorf("failed to list nodes: %w", err)
+	}
+
+	// Create a set of preferred node names for fast lookup
+	preferredNodeSet := sets.New(wObj.Resource.PreferredNodes...)
+
+	// Filter nodes that are in the preferred nodes list and are ready
+	availableBYONodes := make([]*v1.Node, 0, len(nodeList.Items))
+	for i := range nodeList.Items {
+		node := &nodeList.Items[i]
+
+		// if node provision is disabled, preferred nodes will be ignored.
+		if !featuregates.FeatureGates[consts.FeatureFlagDisableNodeAutoProvisioning] {
+			// Check if this node is in the preferred nodes list
+			if !preferredNodeSet.Has(node.Name) {
+				continue
+			}
+		}
+
+		// Check if the node is ready
+		if !IsNodeReady(node) {
+			klog.V(4).InfoS("BYO node is not ready, skipping",
+				"node", node.Name,
+				"workspace", klog.KObj(wObj))
+			continue
+		}
+
+		availableBYONodes = append(availableBYONodes, node)
```

Comment:
Per existing logic, we should also check if there is `nvidia.com/gpu` resource for the node right? https://github.com/kaito-project/kaito/blob/main/pkg/workspace/controllers/workspace_controller.go#L618

Reply:
the above login is extracted as node resource manager, the pull request link: https://github.com/kaito-project/kaito/pull/1418

then nodeclaim manager need to manage nodeclaim resource only.